File: Concordance_Java/src/concordance/PrefixTree.java
/**********************************************
 * Prefix_tree:                               *
 * this class implements a Prefix tree.       *
 * The tree starts from the root, a node      *
 * representing an empty string or content).  *
 * Then, nodes are pushed down the tree       *
 * with the method "add_content".             *
 * Nodes are inserted by insertions and/or    *
 * splitting.                                 *
 * Each node holds a linked list of sorted    *
 * children (i.e., ordered using the ordering *
 * relation defined on the contents).         *
 * Visiting the tree (e.g., using a visitor), *
 * the content of all the tree is printed     *
 * following the sorting of the nodes.        *
 * @note: all operations (insertion, lookup,  *
 * etc.) are performed in O(k) time where k   *
 * is the maximum length of the strings in    *
 * the set. Visiting the tree takes time      *
 * O(nk), where n is the number of the        *
 * strings in the set. Space is O(nk).        *
 **********************************************/
package concordance;

class PrefixTree extends AbstractDataStore {
  
  private int nodeCounter;
  private String dbg;
  /* Root of the tree: a node*/
  private Node root;
  
  private void init() {
   dbg  = "#log: Prefix_tree - ";
   root = new Node();
   nodeCounter = 1;
  }//init
  
  private void visit_tree() {
    if ( verbose ) {
      System.out.println( dbg + "Start visiting PrefixTree");
      System.out.println( dbg + "Creatinr a visitor...");
    }
    /* Create a visitor */
    ConcordanceVisitor concordanceVisitor = 
            new ConcordanceVisitor ( outFile );
    
    if ( verbose ) {
      System.out.println( dbg + "...visitor created");
      System.out.println( dbg + "Visiting PrefixTree...");
      System.out.println( "---------------- CONCORDANCE ------------------");
    }
    
    if ( timer ) setTime();
    
    root.accept( concordanceVisitor );
    
    /* Print to file */
    if ( !outFile.equals( "" ) ) {
      concordanceVisitor.printToFile ();
    }
    
    if ( verbose ) {
      System.out.println( "-----------------------------------------------" );
    }
    if ( timer ) {
      System.out.print( dbg + "Time to visit PrefixTree: ");
      System.out.println( getTime() + "ms" );
    }
  }//visit_tree
  
  
  public PrefixTree () {
    init();
  }//PrefixTree
  
  /* Size of the tree:
   * number of nodes in the tree. */
  @Override
  public int getSize() {
    return nodeCounter;
  }//getSize
  
  /* Add content to the tree.
   * It pushes down the content starting from
   * the root of the tree. */
  @Override
  public void addContent( Content content ) {
    if ( content != null ) {
      /* System.out.println ( (String)(content.getRefBaseObj()) ); */
      root.addContent ( content );
    }
  }//addContent
  
  /* Fill Prefix tree using a parser */
  @Override
  void fillStore ( Parser prs ) {
    
    if ( verbose ) {
      System.out.println( dbg + "Start filling store...");
    }
    
    if ( timer ) setTime();
    
    root.addContent( new Token("") );
    while ( prs.hasMoreElements() ) {
      addContent ( prs.getNextElement() );
    }
    
    if ( timer ) {
      System.out.print( dbg + "Time to fill the tree: ");
      System.out.println( getTime() + "ms" );
    }
    
    if ( verbose ) {
      System.out.println( dbg + "...Filled store.");
    }
    
  }//fillStore
  
  /* Visit tree (it prints the sorted list of
   * the contents of the nodes). */
  @Override
  public void print() {
    visit_tree();
  }//print
  
  /* Prints data structure.
   * In this case, it prints the tree, 
   * i.e., nodes info. */
  @Override
  public void print_data_store () {
    root.print();
  }//print_data_store

}//PrefixTree

File: Concordance_C++/host_core/prefix_tree.cpp
#include "globals.h"
#include "prefix_tree.h"
#include "concordance_visitor.h"
#include "token.h"

//#define PREFIX_TREE_DBG

Prefix_tree::Prefix_tree () {
  _dbg  = "#log: Prefix_tree - ";
  _root = new Node();
}//-

//@todo: use shared_ptr
Prefix_tree::~Prefix_tree () {
  delete _root;
  _root = nullptr;
}//-

int
Prefix_tree::get_size () {
  return _root->get_glb_id();
}//get_size

void
Prefix_tree::add_content ( Content* content ) {
  if ( content != nullptr ) {
#ifdef PREFIX_TREE_DBG
    content->print();
#endif
    _root->add_content ( content );
  }
}//add_content

void
Prefix_tree::print_data_store () {
  _root->print();
}//print

void
Prefix_tree::print () {
  visit_tree();
}//print

void
Prefix_tree::visit_tree () {
  
  if ( _verbose ) {
    std::cout << _dbg << "Start visiting tree\n";
    std::cout << _dbg << "Create visitor...\n";
  }
  
  /// Create a visitor
  ConcordanceVisitor * concordance_visitor =
  new ConcordanceVisitor ( _out_file );
  
  if ( _timer ) set_timer();
  
  if ( _verbose ) {
    std::cout << _dbg << "...visitor created\n";
    std::cout << _dbg << "Visiting tree...\n";
    std::cout << std::endl;
    std::cout << "---------------- CONCORDANCE ------------------\n";
  }
  
  _root->accept ( *concordance_visitor );
  
  if ( _verbose ) {
    std::cout << "-----------------------------------------------\n";
    std::cout << std::endl;
  }
  
  double tot_time;
  if ( _timer ) tot_time = get_timer();
  
  /// Print to file
  if ( _out_file.compare( "" ) != 0 ) {
    concordance_visitor->print_to_file ();
  }
  
  if ( _timer ) {
    std::cout << _dbg << "Time for visiting the Prefix tree: ";
    std::cout <<  tot_time << " sec.\n";
  }
  
  if ( _verbose ) {
    std::cout << _dbg << "...tree visited\n";
    std::cout << _dbg << "Delete visitor...\n";
  }
  
  /// Delete visitor
  delete concordance_visitor;
  
  if ( _verbose ) {
    std::cout << _dbg << "visitor deleted\n";
  }
}//visit_tree

void
Prefix_tree::fill_store ( Parser * parser ) {
  if ( _verbose ) {
    std::cout << _dbg << "Start filling store...\n";
  }
  
  if ( _timer ) set_timer();
  
  /// Add the root -> empty string
  _root->add_content ( new Token("") );
  /*
   * Invoke parser to get all the
   * contents to add (on-line) into the tree.
   */
  while ( parser->has_more_elements() ) {
    add_content ( parser->get_next_content() );
  }
  
  if ( _timer ) {
    std::cout << _dbg << "Time for filling the Prefix tree: ";
    std::cout << get_timer() << " sec.\n";
  }
  
  if ( _verbose ) {
    std::cout << _dbg << "...Filled store.\n";
  }
}//fill_store
